File: mirp.py
import argparse
import os
import shutil
import tempfile
import zipfile


def list_and_extract_zip_contents(tag, zip_file_path, extract_to):
    # Check if the file is a zip file
    if zipfile.is_zipfile(zip_file_path):
        try:
            with zipfile.ZipFile(zip_file_path, 'r') as zip_ref:
                # List all the contents of the zip file
                print(f"Contents of {zip_file_path}:")
                for file_info in zip_ref.infolist():
                    print(f"  {file_info.filename} - {file_info.file_size} bytes")
                    # Skip directories
                    if not file_info.is_dir():
                        # Create a subfolder based on the file's name (excluding the extension)
                        subfolder_name = os.path.splitext(file_info.filename)[0]
                        subfolder_path = os.path.join(extract_to, subfolder_name)
                        os.makedirs(subfolder_path, exist_ok=True)
                        # Create a temporary directory for extraction
                        with tempfile.TemporaryDirectory() as temp_dir:
                            # Extract the file to the temporary directory
                            extracted_path = zip_ref.extract(file_info, temp_dir)
                            # Move the file to the subfolder and rename it
                            file_extension = os.path.splitext(file_info.filename)[1]
                            new_file_name = f"{tag}{file_extension}"
                            new_file_path = os.path.join(subfolder_path, new_file_name)
                            shutil.move(extracted_path, new_file_path)
                print(f"All files extracted to {extract_to}")
        except zipfile.BadZipFile:
            print(f"Error: {zip_file_path} is not a valid zip file.")
        except Exception as e:
            print(f"Error processing {zip_file_path}: {e}")
    else:
        print(f"Error: {zip_file_path} is not a zip file.")

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Process a zip file, list its contents, and extract them.")
    parser.add_argument("tag", help="Tag to rename the extracted files to")
    parser.add_argument("zip_file_path", help="Path to the zip file to be processed")
    parser.add_argument("extract_to", help="Directory to extract the contents to")

    args = parser.parse_args()
    list_and_extract_zip_contents(args.tag, args.zip_file_path, args.extract_to)

File: cmd/mirp.go
package main

import (
	"archive/zip"
	"flag"
	"fmt"
	"io"
	"log"
	"os"
	"path/filepath"
	"strings"
	"sync"

	"github.com/schollz/progressbar/v3"
)

// ExtractFile extracts a single file from the zip archive
func ExtractFile(zipFilePath string, file *zip.File, tag string, extractTo string, wg *sync.WaitGroup, bar *progressbar.ProgressBar) {
	defer wg.Done()

	// Create a subfolder based on the file's name (excluding the extension)
	subfolderName := strings.TrimSuffix(file.Name, filepath.Ext(file.Name))
	subfolderPath := filepath.Join(extractTo, subfolderName)
	err := os.MkdirAll(subfolderPath, os.ModePerm)
	if err != nil {
		log.Printf("Error creating directory %s: %v", subfolderPath, err)
		return
	}

	// Open the file inside the zip archive
	rc, err := file.Open()
	if err != nil {
		log.Printf("Error opening file %s: %v", file.Name, err)
		return
	}
	defer rc.Close()

	// Determine the new file path
	fileExtension := filepath.Ext(file.Name)
	newFileName := fmt.Sprintf("%s%s", tag, fileExtension)
	newFilePath := filepath.Join(subfolderPath, newFileName)

	// Create the new file
	newFile, err := os.Create(newFilePath)
	if err != nil {
		log.Printf("Error creating file %s: %v", newFilePath, err)
		return
	}
	defer newFile.Close()

	// Copy the contents to the new file
	_, err = io.Copy(newFile, rc)
	if err != nil {
		log.Printf("Error writing to file %s: %v", newFilePath, err)
		return
	}

	bar.Add(1)
}

// ListAndExtractZipContents lists and extracts the contents of the zip file
func ListAndExtractZipContents(tag string, zipFilePath string, extractTo string) {
	r, err := zip.OpenReader(zipFilePath)
	if err != nil {
		log.Fatalf("Error opening zip file %s: %v", zipFilePath, err)
	}
	defer r.Close()

	var wg sync.WaitGroup
	bar := progressbar.Default(int64(len(r.File)))

	for _, file := range r.File {
		if !file.FileInfo().IsDir() {
			wg.Add(1)
			go ExtractFile(zipFilePath, file, tag, extractTo, &wg, bar)
		}
	}

	wg.Wait()
	fmt.Printf("All files extracted to %s\n", extractTo)
}

func main() {
  fmt.Println("MIRP: Minecraft Resource Pack Indexer")

	tag := flag.String("tag", "", "Tag to rename the extracted files to")
	zipFilePath := flag.String("zip_file_path", "", "Path to the zip file to be processed")
	extractTo := flag.String("extract_to", "", "Directory to extract the contents to")

	flag.Parse()

	if *tag == "" || *zipFilePath == "" || *extractTo == "" {
		log.Fatal("All arguments (tag, zip_file_path, extract_to) are required")
	}

	ListAndExtractZipContents(*tag, *zipFilePath, *extractTo)
}
